Source/WebCore/dom/StaticRange.cpp

```diff
@@ -83,4 +83,31 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
     addWebCoreOpaqueRoot(visitor, end.container.get());
 }
 
+bool StaticRange::isValid() const
+{
+    Node& startContainer = this->startContainer();
+    Node& endContainer = this->endContainer();
+
+    // Its start is before or equal to its end.
+    if (endOffset() < startOffset() && &startContainer == &endContainer)
+        return false;
+    // Its start and end are in the same node tree.
+    if (!connectedInSameTreeScope(&startContainer.rootNode(), &endContainer.rootNode()))
+        return false;
+    // Its start offset is between 0 and its start node’s length, inclusive.
+    // Its end offset is between 0 and its end node’s length, inclusive.
+    return endOffset() <= endContainer.length() && startOffset() <= startContainer.length();
+}
+
+bool StaticRange::computeValidityForPaintingHighlights() const
+{
+    Node& startContainer = this->startContainer();
+    Node& endContainer = this->endContainer();
+
+    if (collapsed())
+        return false;
+    if (is_gt(treeOrder<ComposedTree>(startContainer, endContainer)))
```

Comment:
This check should probably be a part of StaticRange::isValid since the definition of being valid includes that condition:
https://dom.spec.whatwg.org/#staticrange-valid

Reply:
The `if (is_gt(treeOrder<ComposedTree>(startContainer, endContainer)))`?

Reply:
Yes. "Its [start](https://dom.spec.whatwg.org/#concept-range-start) is [before](https://dom.spec.whatwg.org/#concept-range-bp-before) or [equal](https://dom.spec.whatwg.org/#concept-range-bp-equal) to its [end](https://dom.spec.whatwg.org/#concept-range-end)." corresponds to that check.